ReSharper.FSharp/src/FSharp/FSharp.Psi.Intentions/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -162,42 +207,72 @@ module SpecifyTypes =
+            let listConstPat = (ParenPatUtil.addParens listConsParenPat).As<IListConsPat>()
+            let typedPat = (FSharpPatternUtil.getLastTailPattern listConstPat).As<ITypedPat>()
+            (typedPat: ITreeNode), fcsType, typedPat.TypeUsage
+
+        else (typedPat: ITreeNode), fcsType, typedPat.TypeUsage
 
     let private specifyParameterTypes (decl: IParameterOwnerMemberDeclaration) (mfv: FSharpMemberOrFunctionOrValue) displayContext =
-        let rec specifyParameterTypes (fcsParamGroups: 'a seq) (getFcsType: 'a -> FSharpType) (enumerate: 'a -> 'a seq) (parameters: IFSharpPattern seq) isTopLevel =
-            for fcsParamGroup, parameter in Seq.zip fcsParamGroups parameters do
+        let rec specifyParameterTypes (fcsParamGroups: 'a seq) (getFcsType: 'a -> FSharpType) (enumerate: 'a -> 'a seq) (parameters: IFSharpPattern seq) acc isTopLevel =
+            Seq.zip fcsParamGroups parameters
+            |> Seq.fold (fun acc (fcsParamGroup, parameter) ->
                 match parameter.IgnoreInnerParens() with
-                | :? IConstPat | :? ITypedPat -> ()
+                | :? IConstPat | :? ITypedPat -> acc
                 | TupleLikePattern pat when isTopLevel ->
                     let fcsTypes = enumerate fcsParamGroup
-                    specifyParameterTypes fcsTypes getFcsType enumerate pat.Patterns false
+                    specifyParameterTypes fcsTypes getFcsType enumerate pat.Patterns acc false
                 | pattern ->
                     let fcsType = getFcsType fcsParamGroup
-                    specifyPatternType displayContext fcsType pattern
+                    specifyPatternTypeImpl displayContext fcsType pattern :: acc
+            ) acc
 
         let parameters = decl.ParametersDeclarations |> Seq.map _.Pattern
 
         if mfv.IsMember then
             let enumerate x = x |> Seq.map (fun x -> [|x|] :> IList<_>)
             let types = mfv.CurriedParameterGroups
-            specifyParameterTypes types (fun x -> x[0].Type) enumerate parameters true
+            specifyParameterTypes types (fun x -> x[0].Type) enumerate parameters [] true
         else
             let types = getFunctionTypeArgs true mfv.FullType
-            specifyParameterTypes types id (_.GenericArguments) parameters true
+            specifyParameterTypes types id (_.GenericArguments) parameters [] true
+
+    let private bindAnnotations annotationsInfo =
+        let annotationsInfo =
+            annotationsInfo
+            |> Seq.fold collectTypeUsages []
+            |> Seq.toList
+
+        match annotationsInfo with
+        | [] -> ()
+        | (typeUsage, _, _) :: _ ->
+
+        use pinResultsCookie = typeUsage.FSharpFile.PinTypeCheckResults(true, "Specify types")
+
+        for typeUsage, fcsType, context: ITreeNode in annotationsInfo do
+            let typeReference = typeUsage.ReferenceName
+            let reference = typeReference.Reference.AllowAllSymbolCandidatesCheck()
```

Comment:
Do we still need it in this PR?

Reply:
This is necessary for cases, for example, when `string` resolves to both the type and the `string` operator (which appears first in a symbol’s list).